dxcint/dxcint/Context.py

```diff
@@ -0,0 +1,153 @@
+import dxpy
+import inspect
+import os
+import re
+from threading import Lock
+from typing import Optional
+from dxpy.api import project_new_folder, project_describe, project_remove_folder
+
+from dxcint.Logger import Logger
+
+
+class ContextError(Exception):
+    """
+    Class to handle Context errors
+    """
+
+
+class Context(object):
+    def __init__(
+            self,
+            project: str,
+            repo_root: str,
+            folder: Optional[str] = None,
+            logger_verbosity: str = "info"
+    ):
+        self._logger = Logger.make(name=__name__, verbosity=logger_verbosity)
+        self._project_id = self._resolve_project(project=project)
+        self._user = dxpy.whoami()
+        self._repo_root_dir = repo_root
+        self._compiler_version = self._get_version()
+        self._platform_build_dir = self._create_platform_build_folder(folder)
+        self._lock = Lock()
+        self._project_info = project_describe(self._project_id)
+
+    def __setattr__(self, *args):
+        if inspect.stack()[1][3] == "__init__":
+            object.__setattr__(self, *args)
+        else:
+            raise ContextError("Context class is immutable")
+
+    @property
+    def project_id(self):
+        return self._project_id
+
+    @property
+    def repo_root_dir(self):
+        return self._repo_root_dir
+
+    @property
+    def platform_build_dir(self):
+        return self._platform_build_dir
+
+    @property
+    def version(self):
+        return self._compiler_version
+
+    @property
+    def lock(self):
+        return self._lock
+
+    @property
+    def project_info(self):
+        return self._project_info
+
+    @property
+    def logger(self):
+        return self._logger
+
+    @staticmethod
+    def _resolve_project(project: str) -> str:
+        # First, see if the project is a project-id.
+        try:
+            project = dxpy.DXProject(project)
+            return project.get_id()
```

Comment:
I think it is better not to reuse the`project` var name. maybe use `found_project` instead?

Reply:
thanks, renamed in [cb81b09](https://github.com/dnanexus/dxCompiler/pull/312/commits/cb81b094cc5073b769b829ca017d221a92fb4b99)